### ISSUE TITLE
Record definition doesn't compile without a specific `let` binding

### ISSUE THREAD

**Issue by bgavran** [agda/agda#4131]
The following function is a part of a record:

```agda
  αₘ : {a b c : obj}
    → ((a ⊗ₒ b) ⊗ₒ c)
    hom (a ⊗ₒ (b ⊗ₒ c))
  αₘ = η (forward associator)
```
and the entire record fails to compile unless I change the function implementation to:
```agda
  αₘ : {a b c : obj}
    → ((a ⊗ₒ b) ⊗ₒ c)
    hom (a ⊗ₒ(b ⊗ₒ c))
  αₘ = let t =  η (forward associator) in t
```

The error which shows up when compiling the record with the first function is:
```
obj !=< Σ obj (λ x → Cat.obj (cat X cat)) of type Set n
when checking that the expression η (forward associator) has type
{a b c : obj} →
curry (mapObj ⊗) (curry (mapObj ⊗) a b) c hom
curry (mapObj ⊗) a (curry (mapObj ⊗) b c)
```

I would expect not to have to write that `let` binding and I'm not sure how to approach the issue.
The only code this particular problematic code is the following field, which seems to be related to the issue, as the issue doesn't happen without this field.
```agda
  field
    ▵-identity : {a c : obj}
      → αₘ {a = a} {b = 𝟙} {c = c} ● (id ⊗ₘ λₘ) ≡ ρₘ ⊗ₘ id
```

The entire code is part of my larger project where I'm formalizing CT. I've managed to remove all the extra code that isn't related to the error and I've made a repo of it [here](https://github.com/bgavran/Agda_CT_Bug).

The file with the problematic code is `Monoidal.agda` where I'm defining a monoidal category. The issue seems to be related to the interspersal of fields and regular definitions, but as far as I understand, these definitions should compile okay.

**Comment by laMudri** [comment]
I've cut it down this far:

```agda
module Category where

open import Data.Product
open import Data.Unit
open import Function renaming (id to id→; _∘_ to _∘ᶠ_)

record Cat : Set₁ where
  constructor MkCat

  field
    obj : Set
    _hom_ : (a b : obj) → Set

record _Functor_ (cat1 : Cat) (cat2 : Cat) : Set where
  constructor MkFunctor
  open Cat
  field
    mapObj : obj cat1 -> obj cat2

idFunctor : {cat : Cat} -> cat Functor cat
idFunctor = record { mapObj = id→ }


_∘F_ : ∀ {cat1 : Cat} -> {cat2 : Cat} -> {cat3 : Cat}
  -> (g :              cat2 Functor cat3)
  -> (f : cat1 Functor cat2)
  -> (    cat1        Functor       cat3)
_∘F_ (MkFunctor mapObj₂) (MkFunctor mapObj₁) = MkFunctor (mapObj₂ ∘ᶠ mapObj₁)

module _ {cat1 : Cat} {cat2 : Cat} where

  open Cat

  record _NatTrans_ (fun1 fun2 : cat1 Functor cat2) : Set where
    constructor MkNatTrans
    open _Functor_ fun1 renaming (mapObj to mapObj₁)
    open _Functor_ fun2 renaming (mapObj to mapObj₂)
    field
      η : {a : obj cat1} → _hom_ cat2 (mapObj₁ a) (mapObj₂ a)

  -- actually should be called naturalTransformation category
  functorCategory : Cat
  functorCategory = MkCat (cat1 Functor cat2) _NatTrans_

module _ {cat : Cat} where

  open Cat cat

  record _≅_ (a : obj) (b : obj) : Set where
    constructor MkIso
    field
      forward : a hom b

  open _≅_

module _ where

  open Cat
  open _Functor_

  private
    variable
      c₁ c₂ c₃ d₁ d₂ : Cat

  _X_ : Cat → Cat → Cat
  obj (c₁ X c₂) = (obj c₁ × obj c₂)
  _hom_ (c₁ X c₂) (a₁ , a₂) (b₁ , b₂) = (a₁ hom₁ b₁) × (a₂ hom₂ b₂)
    where _hom₁_ = _hom_ c₁
          _hom₂_ = _hom_ c₂

  productAssociatorᵣ : (c₁ X (c₂ X c₃)) Functor ((c₁ X c₂) X c₃)
  productAssociatorᵣ = MkFunctor
    < < proj₁ , proj₁ ∘ᶠ proj₂ > , proj₂ ∘ᶠ proj₂ >

  _𝕏_ : (c₁ Functor d₁) → (c₂ Functor d₂) → (c₁ X c₂) Functor (d₁ X d₂)
  mapObj (F 𝕏 G) (a , a') = mapObj F a , mapObj G a'

module Monoidal (cat : Cat) where

  open Cat cat
  open _≅_
  open _Functor_
  open _NatTrans_

  record Monoidal : Set where
    constructor MkMonoidal

    field
      ⊗ : (cat X cat) Functor cat

    x⊗[y⊗z] : (cat X (cat X cat)) Functor cat
    x⊗[y⊗z] = ⊗ ∘F (idFunctor 𝕏 ⊗)

    [x⊗y]⊗z : (cat X (cat X cat)) Functor cat
    [x⊗y]⊗z = ⊗ ∘F ((⊗ 𝕏 idFunctor {cat = cat}) ∘F productAssociatorᵣ)

    field
      associator  : _≅_ {cat = functorCategory} [x⊗y]⊗z x⊗[y⊗z]

    infixl 10 _⊗ₒ_
    _⊗ₒ_ : obj → obj → obj
    _⊗ₒ_ = curry (mapObj ⊗)

    αₘ : {a b c : obj}
      → ((a ⊗ₒ b) ⊗ₒ c)
      hom (a ⊗ₒ(b ⊗ₒ c))
    -- αₘ = η (forward associator)
    αₘ = let t = η (forward associator) in t

    field
      ▵-identity : ⊤
```

Not sure what the problem is still, but it has something to do with how definitions before fields are handled.

**Comment by guillaumebrunerie** [comment]
Here is an even shorter example. The idea is that we should have `f' {x} {y} = f {(x , y)}` but Agda only figures it out when using `f' = let u = f in u`.
It seems to have to do with the fact that the let in for `f'` does not maximally insert its implicit arguments, but the one for `u` does.

```agda
open import Data.Product

postulate
  A : Set
  P : A → A → Set
  f : {x : A × A} → P (proj₁ x) (proj₂ x)

test : Set
test =
  let f' : {x y : A} → P x y
--    f' = f               -- Fails
      f' = let u = f in u  -- Works
  in {!!}
```

**Comment by guillaumebrunerie** [comment]
Definitions before fields are translated to let-in expressions, and this creates various issues. See #434.

**Comment by guillaumebrunerie** [comment]
A very similar issue is #3381.

**Comment by bgavran** [comment]
Thanks for making the example more precise, and for linking this to other issues. Putting an empty pattern match `{_}` on  `αₘ` seems to mitigate the issue, for now.

**Comment by UlfNorell** [comment]
Further simplified:
```agda
postulate
  A : Set

works : Set
works = let B : {X : Set} → Set
            B = let _ = Set in A
        in A

fails : Set
fails = let B : {X : Set} → Set
            B = A    -- Set !=< {X : Set} → Set
        in A
```

**Comment by UlfNorell** [comment]
This was introduced by my fix of #655. That changed let definitions to be checked as aliases. The problem is that this applies also when there is a type signature. In this case you do want implicit argument insertion to kick in.
